### ISSUE TITLE
Enable parallel testing with @test and @testset

### ISSUE THREAD

**Issue by danielsoutar** [JuliaLang/julia#43662]
Hi there, wonderful language you guys have and look forward to engaging further with the community.

I am working on a production codebase that has a sizeable number of tests, and we expect this to grow significantly over time. As such, it is highly desirable to accelerate the tests as well as some of the more mathematical code we have.

I’ve noticed that while Julia has an impressive variety of parallel constructs, they don’t map well to testing. For instance, the `@testset` being a macro makes it more awkward to compose, say by wrapping `testset`s up in an array and then running `Threads.@threads` over them in a for-loop. You can wrap these `testset`s in functions, but then the issue is the enclosing `testset` doesn't have access to them and you lose the summary. In both cases, the force-printing to stdout means that you quickly run into gibberish and have no way of controlling this.

The following might explain this better:

```julia
danielsoutar@MBP danielsoutar % julia --threads 4
julia> Threads.nthreads()
4
julia> using Test
julia> foo(x) = length(x)^2
foo (generic function with 1 method)
julia> a() = begin
           @testset "trigonometric identities" begin
               θ = 2/3*π
               @test sin(-θ) ≈ -sin(θ)
               @test cos(-θ) ≈ cos(θ)
               @test sin(2θ) ≈ 2*sin(θ)*cos(θ)
               @test cos(2θ) ≈ cos(θ)^2 - sin(θ)^2
           end
       end;
julia> b() = begin
           @testset "Foo Tests---------------" begin
               @test foo("a")   == 1
               @test foo("ab")  == 4
               @test foo("abc") == 9
           end;
       end;
julia> tests = [a, a, a, a, b, b, b, b];
julia> @testset verbose=false "enclosing" begin
           Threads.@threads for i in 1:length(tests)
               tests[i]()
           end
       end;
Test Summary:            | Test Summary:            | Test Summary:            | Test Summary:            | Pass  Pass  Pass  Pass  TotalTotalTotalTotal



trigonometric identities | trigonometric identities   4      4 | Foo Tests---------------   4   | 
    4   3  Foo Tests---------------
 |     3Test Summary:            |    3  
Test Summary:            | Pass      3Pass  
TotalTest Summary:            | Total

Pass  trigonometric identitiesTest Summary:            |  | Totaltrigonometric identities   4   | 
Pass  Foo Tests---------------    4   4  Total | 

    4Foo Tests---------------   3   | 
   3      3
    3
Test Summary: |
enclosing     | No tests
```

I'm not sure if creating a new type of `TestSet` would work here. While I'm no expert at programming languages, to me the problem is best solved at the language level and either providing a primitive like a `parallel_testset` macro that can execute its testsets in parallel according to the number of specified threads, or redesigning `testset`. The latter would (I'm guessing) mean making printing configurable, disabling it when parallelism is desired, and then combining the results of the returned testsets and printing them as a whole in the enclosing testset. Having to use an external library runs the risk of having several `testset` macros with inconsistent functionality and semantics in the codebase. Especially since all of those libraries I've seen are very young and changing rapidly with the language, so this is even less desirable.

FWIW I'd be happy to help contribute towards something like this as this would be really useful to me and my team. Thanks for your time.

**Comment by DilumAluthge** [comment]
There are several packages in the ecosystem that provide parallel testing, e.g. https://github.com/RelationalAI-oss/XUnit.jl

**Comment by Sacha0** [comment]
https://github.com/JuliaTesting/ReTest.jl also offers parallel testing and has more community support than XUnit.jl (which regrettably has no maintainer at this point).

**Comment by vtjnash** [comment]
Is it worthwhile to convert this to a discussion?

**Comment by danielsoutar** [comment]
> https://github.com/JuliaTesting/ReTest.jl also offers parallel testing and has more community support than XUnit.jl (which regrettably has no maintainer at this point).

I think it is worth converting this to a discussion. On your recommendation I tried ReTest, which in addition to some limitations that I ran into (and opened several issues), it seems it is not actively maintained either. The last time the maintainer engaged at all with the repo via commenting on issues was nearly 8 months ago and there's been no commits for nearly a year now.
